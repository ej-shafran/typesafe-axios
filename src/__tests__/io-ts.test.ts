import * as t from "io-ts";
import { isLeft, isRight } from "fp-ts/Either";
import nock from "nock";
import { taxios } from "..";
import * as UTILS from "./__utils__";

const SCHEMA = t.type({
  id: t.string,
  username: t.string,
  email: t.string,
  address: t.intersection([
    t.type({ street: t.string }),
    t.partial({ zipcode: t.string }),
  ]),
  followers: t.array(
    t.intersection([
      t.type({ id: t.string }),
      t.partial({ role: t.union([t.literal("admin"), t.literal("guest")]) }),
    ])
  ),
});

type TYPE = t.TypeOf<typeof SCHEMA>;
function ensure(_: TYPE) { }

describe("io-ts interop", () => {
  it("returns a Right with the data if the data is valid", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.VALID);

    const { data } = await taxios(SCHEMA.decode).get(UTILS.API);

    expect(isRight(data)).toBe(true);
    if (isRight(data)) {
      expect(data.right).toEqual(UTILS.VALID);
      ensure(data.right);
    }
  });

  it("returns a Left with the errors if the data is invalid", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.INVALID);

    const { data } = await taxios(SCHEMA.decode).get(UTILS.API);

    expect(isLeft(data)).toBe(true);
    if (isLeft(data)) {
      expect(data.left).toMatchSnapshot();
    }
  });
});
