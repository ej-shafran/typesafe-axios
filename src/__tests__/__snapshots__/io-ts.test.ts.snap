// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`io-ts interop returns a Left with the errors if the data is invalid 1`] = `
[
  {
    "context": [
      {
        "actual": {
          "address": "ABC123 St.",
          "followers": [],
          "id": "SOME_ID",
          "username": "SOME_USERNAME",
        },
        "key": "",
        "type": InterfaceType {
          "_tag": "InterfaceType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "{ id: string, username: string, email: string, address: ({ street: string } & Partial<{ zipcode: string }>), followers: Array<({ id: string } & Partial<{ role: ("admin" | "guest") }>)> }",
          "props": {
            "address": IntersectionType {
              "_tag": "IntersectionType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "({ street: string } & Partial<{ zipcode: string }>)",
              "types": [
                InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ street: string }",
                  "props": {
                    "street": StringType {
                      "_tag": "StringType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "string",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
                PartialType {
                  "_tag": "PartialType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "Partial<{ zipcode: string }>",
                  "props": {
                    "zipcode": StringType {
                      "_tag": "StringType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "string",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
              ],
              "validate": [Function],
            },
            "email": StringType {
              "_tag": "StringType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "string",
              "validate": [Function],
            },
            "followers": ArrayType {
              "_tag": "ArrayType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "Array<({ id: string } & Partial<{ role: ("admin" | "guest") }>)>",
              "type": IntersectionType {
                "_tag": "IntersectionType",
                "decode": [Function],
                "encode": [Function],
                "is": [Function],
                "name": "({ id: string } & Partial<{ role: ("admin" | "guest") }>)",
                "types": [
                  InterfaceType {
                    "_tag": "InterfaceType",
                    "decode": [Function],
                    "encode": [Function],
                    "is": [Function],
                    "name": "{ id: string }",
                    "props": {
                      "id": StringType {
                        "_tag": "StringType",
                        "decode": [Function],
                        "encode": [Function],
                        "is": [Function],
                        "name": "string",
                        "validate": [Function],
                      },
                    },
                    "validate": [Function],
                  },
                  PartialType {
                    "_tag": "PartialType",
                    "decode": [Function],
                    "encode": [Function],
                    "is": [Function],
                    "name": "Partial<{ role: ("admin" | "guest") }>",
                    "props": {
                      "role": UnionType {
                        "_tag": "UnionType",
                        "decode": [Function],
                        "encode": [Function],
                        "is": [Function],
                        "name": "("admin" | "guest")",
                        "types": [
                          LiteralType {
                            "_tag": "LiteralType",
                            "decode": [Function],
                            "encode": [Function],
                            "is": [Function],
                            "name": ""admin"",
                            "validate": [Function],
                            "value": "admin",
                          },
                          LiteralType {
                            "_tag": "LiteralType",
                            "decode": [Function],
                            "encode": [Function],
                            "is": [Function],
                            "name": ""guest"",
                            "validate": [Function],
                            "value": "guest",
                          },
                        ],
                        "validate": [Function],
                      },
                    },
                    "validate": [Function],
                  },
                ],
                "validate": [Function],
              },
              "validate": [Function],
            },
            "id": StringType {
              "_tag": "StringType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "string",
              "validate": [Function],
            },
            "username": StringType {
              "_tag": "StringType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "string",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      {
        "actual": undefined,
        "key": "email",
        "type": StringType {
          "_tag": "StringType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "string",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": undefined,
  },
  {
    "context": [
      {
        "actual": {
          "address": "ABC123 St.",
          "followers": [],
          "id": "SOME_ID",
          "username": "SOME_USERNAME",
        },
        "key": "",
        "type": InterfaceType {
          "_tag": "InterfaceType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "{ id: string, username: string, email: string, address: ({ street: string } & Partial<{ zipcode: string }>), followers: Array<({ id: string } & Partial<{ role: ("admin" | "guest") }>)> }",
          "props": {
            "address": IntersectionType {
              "_tag": "IntersectionType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "({ street: string } & Partial<{ zipcode: string }>)",
              "types": [
                InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ street: string }",
                  "props": {
                    "street": StringType {
                      "_tag": "StringType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "string",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
                PartialType {
                  "_tag": "PartialType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "Partial<{ zipcode: string }>",
                  "props": {
                    "zipcode": StringType {
                      "_tag": "StringType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "string",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
              ],
              "validate": [Function],
            },
            "email": StringType {
              "_tag": "StringType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "string",
              "validate": [Function],
            },
            "followers": ArrayType {
              "_tag": "ArrayType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "Array<({ id: string } & Partial<{ role: ("admin" | "guest") }>)>",
              "type": IntersectionType {
                "_tag": "IntersectionType",
                "decode": [Function],
                "encode": [Function],
                "is": [Function],
                "name": "({ id: string } & Partial<{ role: ("admin" | "guest") }>)",
                "types": [
                  InterfaceType {
                    "_tag": "InterfaceType",
                    "decode": [Function],
                    "encode": [Function],
                    "is": [Function],
                    "name": "{ id: string }",
                    "props": {
                      "id": StringType {
                        "_tag": "StringType",
                        "decode": [Function],
                        "encode": [Function],
                        "is": [Function],
                        "name": "string",
                        "validate": [Function],
                      },
                    },
                    "validate": [Function],
                  },
                  PartialType {
                    "_tag": "PartialType",
                    "decode": [Function],
                    "encode": [Function],
                    "is": [Function],
                    "name": "Partial<{ role: ("admin" | "guest") }>",
                    "props": {
                      "role": UnionType {
                        "_tag": "UnionType",
                        "decode": [Function],
                        "encode": [Function],
                        "is": [Function],
                        "name": "("admin" | "guest")",
                        "types": [
                          LiteralType {
                            "_tag": "LiteralType",
                            "decode": [Function],
                            "encode": [Function],
                            "is": [Function],
                            "name": ""admin"",
                            "validate": [Function],
                            "value": "admin",
                          },
                          LiteralType {
                            "_tag": "LiteralType",
                            "decode": [Function],
                            "encode": [Function],
                            "is": [Function],
                            "name": ""guest"",
                            "validate": [Function],
                            "value": "guest",
                          },
                        ],
                        "validate": [Function],
                      },
                    },
                    "validate": [Function],
                  },
                ],
                "validate": [Function],
              },
              "validate": [Function],
            },
            "id": StringType {
              "_tag": "StringType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "string",
              "validate": [Function],
            },
            "username": StringType {
              "_tag": "StringType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "string",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      {
        "actual": "ABC123 St.",
        "key": "address",
        "type": IntersectionType {
          "_tag": "IntersectionType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "({ street: string } & Partial<{ zipcode: string }>)",
          "types": [
            InterfaceType {
              "_tag": "InterfaceType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "{ street: string }",
              "props": {
                "street": StringType {
                  "_tag": "StringType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "string",
                  "validate": [Function],
                },
              },
              "validate": [Function],
            },
            PartialType {
              "_tag": "PartialType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "Partial<{ zipcode: string }>",
              "props": {
                "zipcode": StringType {
                  "_tag": "StringType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "string",
                  "validate": [Function],
                },
              },
              "validate": [Function],
            },
          ],
          "validate": [Function],
        },
      },
      {
        "actual": "ABC123 St.",
        "key": "0",
        "type": InterfaceType {
          "_tag": "InterfaceType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "{ street: string }",
          "props": {
            "street": StringType {
              "_tag": "StringType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "string",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": "ABC123 St.",
  },
  {
    "context": [
      {
        "actual": {
          "address": "ABC123 St.",
          "followers": [],
          "id": "SOME_ID",
          "username": "SOME_USERNAME",
        },
        "key": "",
        "type": InterfaceType {
          "_tag": "InterfaceType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "{ id: string, username: string, email: string, address: ({ street: string } & Partial<{ zipcode: string }>), followers: Array<({ id: string } & Partial<{ role: ("admin" | "guest") }>)> }",
          "props": {
            "address": IntersectionType {
              "_tag": "IntersectionType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "({ street: string } & Partial<{ zipcode: string }>)",
              "types": [
                InterfaceType {
                  "_tag": "InterfaceType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "{ street: string }",
                  "props": {
                    "street": StringType {
                      "_tag": "StringType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "string",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
                PartialType {
                  "_tag": "PartialType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "Partial<{ zipcode: string }>",
                  "props": {
                    "zipcode": StringType {
                      "_tag": "StringType",
                      "decode": [Function],
                      "encode": [Function],
                      "is": [Function],
                      "name": "string",
                      "validate": [Function],
                    },
                  },
                  "validate": [Function],
                },
              ],
              "validate": [Function],
            },
            "email": StringType {
              "_tag": "StringType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "string",
              "validate": [Function],
            },
            "followers": ArrayType {
              "_tag": "ArrayType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "Array<({ id: string } & Partial<{ role: ("admin" | "guest") }>)>",
              "type": IntersectionType {
                "_tag": "IntersectionType",
                "decode": [Function],
                "encode": [Function],
                "is": [Function],
                "name": "({ id: string } & Partial<{ role: ("admin" | "guest") }>)",
                "types": [
                  InterfaceType {
                    "_tag": "InterfaceType",
                    "decode": [Function],
                    "encode": [Function],
                    "is": [Function],
                    "name": "{ id: string }",
                    "props": {
                      "id": StringType {
                        "_tag": "StringType",
                        "decode": [Function],
                        "encode": [Function],
                        "is": [Function],
                        "name": "string",
                        "validate": [Function],
                      },
                    },
                    "validate": [Function],
                  },
                  PartialType {
                    "_tag": "PartialType",
                    "decode": [Function],
                    "encode": [Function],
                    "is": [Function],
                    "name": "Partial<{ role: ("admin" | "guest") }>",
                    "props": {
                      "role": UnionType {
                        "_tag": "UnionType",
                        "decode": [Function],
                        "encode": [Function],
                        "is": [Function],
                        "name": "("admin" | "guest")",
                        "types": [
                          LiteralType {
                            "_tag": "LiteralType",
                            "decode": [Function],
                            "encode": [Function],
                            "is": [Function],
                            "name": ""admin"",
                            "validate": [Function],
                            "value": "admin",
                          },
                          LiteralType {
                            "_tag": "LiteralType",
                            "decode": [Function],
                            "encode": [Function],
                            "is": [Function],
                            "name": ""guest"",
                            "validate": [Function],
                            "value": "guest",
                          },
                        ],
                        "validate": [Function],
                      },
                    },
                    "validate": [Function],
                  },
                ],
                "validate": [Function],
              },
              "validate": [Function],
            },
            "id": StringType {
              "_tag": "StringType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "string",
              "validate": [Function],
            },
            "username": StringType {
              "_tag": "StringType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "string",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      {
        "actual": "ABC123 St.",
        "key": "address",
        "type": IntersectionType {
          "_tag": "IntersectionType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "({ street: string } & Partial<{ zipcode: string }>)",
          "types": [
            InterfaceType {
              "_tag": "InterfaceType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "{ street: string }",
              "props": {
                "street": StringType {
                  "_tag": "StringType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "string",
                  "validate": [Function],
                },
              },
              "validate": [Function],
            },
            PartialType {
              "_tag": "PartialType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "Partial<{ zipcode: string }>",
              "props": {
                "zipcode": StringType {
                  "_tag": "StringType",
                  "decode": [Function],
                  "encode": [Function],
                  "is": [Function],
                  "name": "string",
                  "validate": [Function],
                },
              },
              "validate": [Function],
            },
          ],
          "validate": [Function],
        },
      },
      {
        "actual": "ABC123 St.",
        "key": "1",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ zipcode: string }>",
          "props": {
            "zipcode": StringType {
              "_tag": "StringType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "string",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": "ABC123 St.",
  },
]
`;
